File: 防火墙项目/firewallclient-nodejs/classes/configDPIRules.js
var protectForm = require('../forms/protectDeviceForm.js');
var fwDeviceForm = require('../forms/fwDeviceForm.js');
var sendInfo = require('../tools/sendInfo.js');
var dbOperations = require('../tools/dbOperations.js');
var configInfo = require('../tools/listenConfigInfo.js');

/**
 * 配置ModbusTCP规则
 * Author：于仁飞
 * @param {*设备对象（防火墙或者被保护设备）} dev 
 * @param {*modbus规则对象} modbusTcpForm 
 * @param {*日志标志} logFlag 
 * @param {*true为添加规则，false为删除规则} addOrDelete 
 */

function configModbusTcpRules(dev, modbusTcpForm, logFlag, addOrDelete) {
    var dpiProto = "modbusTcp";
    var flag;
    var dpiRules, dpiLogRule;
    var sqlRule;

    if (modbusTcpForm.getSrcIp() == "any" && modbusTcpForm.getDstIp() == "any") {
        dpiRules = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + "DROP";

        dpiLogRule = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + "LOG" + " " + "--log-prefix " + "\"" + "DROP&modbus&data_illegal " + "\"";
    }
    if (modbusTcpForm.getSrcIp() == "any" && modbusTcpForm.getDstIp() != "any") {
        dpiRules = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-d" + " " + modbusTcpForm.getDstIp() + " " + "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + "DROP";

        dpiLogRule = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-d" + " " + modbusTcpForm.getDstIp() + " " + "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + "LOG" + " " + "--log-prefix " + "\"" + "DROP&modbus&data_illegal " + "\"";
    }

    if (modbusTcpForm.getSrcIp() != "any" && modbusTcpForm.getDstIp() == "any") {
        dpiRules = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-s " + modbusTcpForm.getSrcIp() + " " + "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + "DROP";

        dpiLogRule = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-s " + modbusTcpForm.getSrcIp() + " " + "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + "LOG" + " " + "--log-prefix " + "\"" + "DROP&modbus&data_illegal " + "\"";
    }

    if (modbusTcpForm.getSrcIp() != "any" && modbusTcpForm.getDstIp() != "any") {
        dpiRules = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-s " + modbusTcpForm.getSrcIp() + " " + "-d" + " " + modbusTcpForm.getDstIp() + " " + "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + " DROP";

        dpiLogRule = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " +
            "-s " + modbusTcpForm.getSrcIp() + " " + "-d" + " " + modbusTcpForm.getDstIp() + " " + "-m" + " " + dpiProto + " " + "--data-addr" + " " + modbusTcpForm.getMinAddr() + ":" + modbusTcpForm.getMaxAddr() + " " + "--modbus-func " +
            modbusTcpForm.getFunc() + " " + "--modbus-data " + modbusTcpForm.getMinData() + ":" + modbusTcpForm.getMaxData() + " -j" + " " + "LOG" + " " + "--log-prefix " + "\"" + "DROP&modbus&data_illegal " + "\"";
    }
    if (addOrDelete) {
        flag = "DPI1";
        sqlRule = "INSERT INTO modbustcp values " + "('" + dev.getDevMAC() + "','" + dpiProto + "','" + modbusTcpForm.getSrcIp() +
            "','" + modbusTcpForm.getDstIp() + "','" + modbusTcpForm.getMinAddr() + "','" + modbusTcpForm.getMaxAddr() + "','" + modbusTcpForm.getMinData() + "','" + modbusTcpForm.getMaxData() + "','" + modbusTcpForm.getFunc() + "','" + logFlag + "','DROP')";
    } else {
        flag = "DPI0";
        sqlRule = "delete from modbustcp where fw_mac= '" + dev.getDevMAC() + "' and protocol = '" + dpiProto + "' and source = '" +
            modbusTcpForm.getSrcIp() + "' and destination = '" + modbusTcpForm.getDstIp() + "' and coiladdressstart = '" + modbusTcpForm.getMinAddr() + "' and coiladdressstart = '" +
            modbusTcpForm.getMaxAddr() + "'and minspeed ='" + modbusTcpForm.getMinData() + "' and maxspeed ='" + modbusTcpForm.getMaxData() + "' and functioncode = '" + modbusTcpForm.getFunc() + "' and log_flag = '" + logFlag + "';";
    }
    var modbusTcpRules = flag + dpiLogRule + " && " + dpiRules;
    sendInfo.sendConfigInfo(dev.getDevIP(), modbusTcpRules);
    dbOperations.dbOperation(sqlRule);
    return (configInfo.listenConfigInfo());
}
exports.configModbusTcpRules = configModbusTcpRules;


/**
 * 用于配置OPC规则
 * Author：于仁飞
 * @param {*防火墙或被保护设备对象} dev 
 * @param {*OPC规则对象} OPCRulesForm 
 * @param {*记录日志} logFlag 
 * @param {*true为添加日志，false为删除日志} addOrDelete 
 */
function configOPCRules(dev, OPCRulesForm, logFlag, addOrDelete) {
    var flag, dpiRules, dpiLogRule, sqlRule;

    if (OPCRulesForm.getSrcIp() == "any" && OPCRulesForm.getDstIp() == "any") {
        dpiRules = "iptables -A FORWARD -p tcp --dport 135 -m state --state ESTABLISHED -j NFQUEUE --queue-num 1";
        dpiLogRule = "iptables -A FORWARD -p tcp --dport 135 -m state --state ESTABLISHED -j LOG --log-prefix " + "\"" + "ACCEPT&OPC&ESTABLISHED " + "\"";
    }
    if (OPCRulesForm.getSrcIp() != "any" && OPCRulesForm.getDstIp() == "any") {
        dpiRules = "iptables -A FORWARD -p tcp -s " + OPCRulesForm.getSrcIp() + " --dport 135 -m state --state ESTABLISHED -j NFQUEUE --queue-num 1";
        dpiLogRule = "iptables -A FORWARD -p tcp -s " + OPCRulesForm.getSrcIp() + " --dport 135 -m state --state ESTABLISHED -j LOG --log-prefix " + "\"" + "ACCEPT&OPC&ESTABLISHED " + "\"";
    }
    if (OPCRulesForm.getSrcIp() == "any" && OPCRulesForm.getDstIp() != "any") {
        dpiRules = "iptables -A FORWARD -p tcp -d " + OPCRulesForm.getDstIp() + " --dport 135 -m state --state ESTABLISHED -j NFQUEUE --queue-num 1";
        dpiLogRule = "iptables -A FORWARD -p tcp -d " + OPCRulesForm.getDstIp() + " --dport 135 -m state --state ESTABLISHED -j LOG --log-prefix " + "\"" + "ACCEPT&OPC&ESTABLISHED " + "\"";
    }
    if (OPCRulesForm.getSrcIp() != "any" && OPCRulesForm.getDstIp() != "any") {
        dpiRules = "iptables -A FORWARD -p tcp -s " + OPCRulesForm.getSrcIp() + " -d " + OPCRulesForm.getDstIp() + " --dport 135 -m state --state ESTABLISHED -j NFQUEUE --queue-num 1";
        dpiLogRule = "iptables -A FORWARD -p tcp -s " + OPCRulesForm.getSrcIp() + " -d " + OPCRulesForm.getDstIp() + " --dport 135 -m state --state ESTABLISHED -j LOG --log-prefix " + "\"" + "ACCEPT&OPC&ESTABLISHED " + "\"";
    }
    if (addOrDelete) {
        flag = "DPI1";
        sqlRule = "INSERT INTO OPC values " + "('" + dev.getDevIP() + "','" + OPCRulesForm.getDstIp() + "','" + OPCRulesForm.getSrcIp() + "','" + logFlag + "','ACCEPT')";
    } else {
        flag = "DPI0";
        sqlRule = "DELETE FROM OPC where (dev_IP='" + dev.getDevIP() + "' and dst_IP='" + OPCRulesForm.getDstIp() + "' and src_IP='" + OPCRulesForm.getSrcIp() + "' and method='ACCEPT')";
    }
    var OPCRules = flag + dpiLogRule + " && " + dpiRules;
    sendInfo.sendConfigInfo(dev.getDevIP(), OPCRules);
    dbOperations.dbOperation(sqlRule);
    return (configInfo.listenConfigInfo());
}
exports.configOPCRules = configOPCRules;


/**
 * 用于配置DNP3规则
 * Author：于仁飞
 * @param {*防火墙或被保护设备对象} dev 
 * @param {*DNP3规则对象} OPCRulesForm 
 * @param {*记录日志} logFlag 
 * @param {*true为添加日志，false为删除日志} addOrDelete 
 */
function configDNP3Rules(dev, DNP3Form, logFlag, addOrDelete) {
    var dpiRules0, dpiRules1, dpiLogRule, flag, sqlRule;
    if (DNP3Form.getSrcIp() == "any" && DNP3Form.getDstIp() == "any") {
        dpiRules0 = "iptables -A FORWARD -p tcp " + " --dport 20000 -m state --state NEW -j DROP";
        dpiRules1 = "iptables -A FORWARD -p tcp " + " --dport 20000 -m state --state ESTABLISHED -j DROP";
        dpiLogRule = "iptables -A FORWARD -p tcp " + " --dport 20000 -m state --state new -j LOG --log-prefix " + "\"" + "DROP&DNP3&illegal " + "\"";
    }
    if (DNP3Form.getSrcIp() != "any" && DNP3Form.getDstIp() == "any") {
        dpiRules0 = "iptables -A FORWARD -p tcp -s " + DNP3Form.getSrcIp() + " --dport 20000 -m state --state NEW -j DROP";
        dpiRules1 = "iptables -A FORWARD -p tcp -s " + DNP3Form.getSrcIp() + " --dport 20000 -m state --state ESTABLISHED -j DROP";
        dpiLogRule = "iptables -A FORWARD -p tcp -s " + DNP3Form.getSrcIp() + " --dport 20000 -m state --state new -j LOG --log-prefix " + "\"" + "DROP&DNP3&illegal " + "\"";
    }
    if (DNP3Form.getSrcIp() == "any" && DNP3Form.getDstIp() != "any") {
        dpiRules0 = "iptables -A FORWARD -p tcp" + " -d " + DNP3Form.getDstIp() + " --dport 20000 -m state --state NEW -j DROP";
        dpiRules1 = "iptables -A FORWARD -p tcp" + " -d " + DNP3Form.getDstIp() + " --dport 20000 -m state --state ESTABLISHED -j DROP";
        dpiLogRule = "iptables -A FORWARD -p tcp" + " -d " + DNP3Form.getDstIp() + " --dport 20000 -m state --state new -j LOG --log-prefix " + "\"" + "DROP&DNP3&illegal " + "\"";
    }
    if (DNP3Form.getSrcIp() != "any" && DNP3Form.getDstIp() != "any") {
        dpiRules0 = "iptables -A FORWARD -p tcp -s " + DNP3Form.getSrcIp() + " -d " + DNP3Form.getDstIp() + " --dport 20000 -m state --state NEW -j DROP";
        dpiRules1 = "iptables -A FORWARD -p tcp -s " + DNP3Form.getSrcIp() + " -d " + DNP3Form.getDstIp() + " --dport 20000 -m state --state ESTABLISHED -j DROP";
        dpiLogRule = "iptables -A FORWARD -p tcp -s " + DNP3Form.getSrcIp() + " -d " + DNP3Form.getDstIp() + " --dport 20000 -m state --state new -j LOG --log-prefix " + "\"" + "DROP&DNP3&illegal " + "\"";
    }
    if (addOrDelete) {
        flag = "DPI1";
        sqlRule = "INSERT INTO dnp3 values " + "('" + dev.getDevIP() + "','" + DNP3Form.getDstIp() + "','" + DNP3Form.getSrcIp() + "','" + logFlag + "','DROP')";
    } else {
        flag = "DPI0";
        sqlRule = "DELETE FROM DNP3 where (dev_IP='" + dev.getDevIP() + "' and dst_IP='" + DNP3Form.getDstIp() + "' and src_IP='" + DNP3Form.getSrcIp() + "' and method='DROP')";
    }
    var DNP3Rules = flag + dpiLogRule + " && " + dpiRules0 + " && " + dpiRules1;
    sendInfo.sendConfigInfo(dev.getDevIP(), DNP3Rules);
    dbOperations.dbOperation(sqlRule);
    return (configInfo.listenConfigInfo());
}
exports.configDNP3Rules = configDNP3Rules;

/**
 * 清除配置的所有规则
 * @param {*防火墙或者被保护设备对象} dev 
 */
function clearAllRules(dev) {
    var clearRule = "iptables -F && iptables -X && iptables -Z && iptables-restore</etc/iptables.up.rules";
    sendInfo.sendConfigInfo(dev.getDevIP(), clearRule);
}
exports.clearAllRules = clearAllRules;

File: 防火墙项目/FWclient-存文件/FWclient/interfaces-classes/ConfigDPIRules.cs


using MySql.Data.MySqlClient;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
namespace FWclient
{
    /// <summary>
    /// 定义防火墙规则
    /// </summary>
    class ConfigDPIRules:IConfigRules
    {
       // private String default_rules = "iptables -P FORWARD  DROP";
        private DeviceForm devform;
        FileOperation of = new FileOperation("C:\\CMP\\Config");
        public ConfigDPIRules(DeviceForm devform)
        {
            this.devform = devform;
        }      

        public bool ConfigModbusTcpRules(ModbusTcpRulesForm mtrf,bool log_flag, bool add_delete)
        {
           // RulesDataProcess.ModbusTcpRulesDataProcess(mtrf);

            String dpi_pro = "modbusTcp";
            string flag=null; string dpi_rules_from_master_to_slave0 = null;
      
            if (mtrf.getSrc_IP() == "any" & mtrf.getDst_IP() == "any")
            {
                dpi_rules_from_master_to_slave0 = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " + "-m" + " " + dpi_pro + " " + "--data-addr" + " " + mtrf.getMin_addr() + ":" + mtrf.getMax_addr() + " " + "--modbus-func " + mtrf.getfunc() + " " + "--modbus-data " + mtrf.getMin_data() + ":" + mtrf.getMax_data() + " -j" + " " + "DROP";
            }
            else if (mtrf.getSrc_IP() == "any" & mtrf.getDst_IP() != "any")
            {
                dpi_rules_from_master_to_slave0 = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " + "-d" + " " + mtrf.getDst_IP() + " " + "-m" + " " + dpi_pro + " " + "--data-addr" + " " + mtrf.getMin_addr() + ":" + mtrf.getMax_addr() + " " + "--modbus-func " + mtrf.getfunc() + " " + "--modbus-data " + mtrf.getMin_data() + ":" + mtrf.getMax_data() + " -j" + " " + "DROP";

            }
            else if (mtrf.getSrc_IP() != "any" & mtrf.getDst_IP() == "any")
            {
                dpi_rules_from_master_to_slave0 = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " + "-s " + mtrf.getSrc_IP() + " "  + "-m" + " " + dpi_pro + " " + "--data-addr" + " " + mtrf.getMin_addr() + ":" + mtrf.getMax_addr() + " " + "--modbus-func " + mtrf.getfunc() + " " + "--modbus-data " + mtrf.getMin_data() + ":" + mtrf.getMax_data() + " -j" + " " + "DROP";
            }
            else
            {

                dpi_rules_from_master_to_slave0 = "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " + "-s " + mtrf.getSrc_IP() + " " + "-d" + " " + mtrf.getDst_IP() + " " + "-m" + " " + dpi_pro + " " + "--data-addr" + " " + mtrf.getMin_addr() + ":" + mtrf.getMax_addr() + " " + "--modbus-func " + mtrf.getfunc() + " " + "--modbus-data " + mtrf.getMin_data() + ":" + mtrf.getMax_data() + " -j" + " " + " DROP";
            }
            //string dpi_rules_from_master_to_slave1 = "iptables -A FORWARD -m state --state ESTABLISHED,RELATED -j ACCEPT";
            string dpi_rules_from_master_to_slave_log= "iptables" + " -A" + " " + "FORWARD" + " " + "-p tcp" + " " + "--dport" + " " + "502" + " " + "-s " + mtrf.getSrc_IP() + " " + "-d" + " " + mtrf.getDst_IP() + " " + "-m" + " " + dpi_pro + " " + "--data-addr" + " " + mtrf.getMin_addr() +":"+ mtrf.getMax_addr() + " " + "--modbus-func " + mtrf.getfunc() + " "+"--modbus-data " + mtrf.getMin_data() + ":" + mtrf.getMax_data() + " -j" + " " + "LOG" + " " + "--log-prefix " + "\"" + "DROP&modbus&data_illegal " + "\"";
           
            if (add_delete == true)
            {
                flag = "DPI1";
           string frule= devform.getDev_IP()+" "+ mtrf.getDst_IP() + " "+ 
                mtrf.getSrc_IP() + " " + mtrf.getMin_addr() + " "+ mtrf.getMax_addr() + " "+ mtrf.getfunc() + " "+ mtrf.getMin_data() + " " + mtrf.getMax_data() + " ACCEPT "+log_flag.ToString();
                of.SaveRules(frule, "modbustcp");
            }

            else if (add_delete == false)
            {
                flag = "DPI0";
                string frule = devform.getDev_IP() + " " +mtrf.getDst_IP() + " " + mtrf.getSrc_IP() + " "+ mtrf.getMin_addr() + 
                   " "+ mtrf.getMax_addr() + " "+ mtrf.getfunc() + " "+ mtrf.getMin_data() + " "+ mtrf.getMax_data() + " ACCEPT "+ log_flag.ToString();
                of.DeleteRules(frule, "modbustcp");
            }
            string rule = flag + dpi_rules_from_master_to_slave_log + " && " + dpi_rules_from_master_to_slave0;
            SendInfo sendcmd = new SendInfo(devform);
            return sendcmd.SendConfigInfo(rule);
        }

        public bool ConfigOPCRules(OPCRulesForm orf,bool log_flag, bool add_delete)
        {
            string flag = null;

            String opc_rules_from_client_to_server0 = "iptables -A FORWARD -p tcp -s " + orf.getSrc_IP() + " -d " + orf.getDst_IP() + " --dport 135 -m state --state ESTABLISHED -j NFQUEUE --queue-num 1";
            String opc_rules_from_client_to_server1 = "iptables -A FORWARD -p tcp -s "+ orf.getDst_IP() +" -d "+ orf.getSrc_IP();
            String opc_rules_from_client_to_server_log= "iptables -A FORWARD -p tcp -s " + orf.getSrc_IP() + " -d " + orf.getDst_IP() + " --dport 135 -m state --state ESTABLISHED -j LOG --log-prefix "+ "\"" + "ACCEPT&OPC&ESTABLISHED " + "\"";
            //String opc_rules_from_server_to_client = "iptables -A FORWARD -p tcp -s " + orf.getDst_IP() + " -d " + orf.getSrc_IP() + " --sport 135 -m state --state ESTABLISHED -j NFQUEUE --queue-num 1";
            if (add_delete == true)
            {
                flag = "DPI1";
                string frule = devform.getDev_IP()+" "+orf.getDst_IP()+" "+orf.getSrc_IP()+" ACCEPT "+log_flag.ToString();
                of.SaveRules(frule, "opc");
            }

            else if (add_delete == false)
            {
                flag = "DPI0";
                string frule = devform.getDev_IP() + " " + orf.getDst_IP() + " " + orf.getSrc_IP() + " ACCEPT " + log_flag.ToString();
                of.DeleteRules(frule, "opc");
            }

            string rule = flag + opc_rules_from_client_to_server_log + " && " + opc_rules_from_client_to_server0 + " && " + opc_rules_from_client_to_server1;
            SendInfo sendcmd = new SendInfo(devform);                  
            return sendcmd.SendConfigInfo(rule);
        }

        public bool ConfigDNP3Rules(DNP3RulesForm dnp3rf,bool log_flag, bool add_delete)
        {
            string flag = null;

            String dnp3_rules_from_client_to_server_new = "iptables -A FORWARD -p tcp -s " + dnp3rf.getSrc_IP() + " -d " + dnp3rf.getDst_IP() + " --dport 20000 -m state --state NEW -j ACCEPT";
           // String dnp3_rules_from_server_to_client_new = "iptables -A FORWARD -p tcp -s " + dnp3rf.getDst_IP() + " -d " + dnp3rf.getSrc_IP() + " --sport 20000 -m state --state NEW -j ACCEPT";
            String dnp3_rules_from_client_to_server_established = "iptables -A FORWARD -p tcp -s " + dnp3rf.getSrc_IP() + " -d " + dnp3rf.getDst_IP() + " --dport 20000 -m state --state ESTABLISHED -j ACCEPT";
            String dnp3_rules_from_client_to_server_back = "iptables -A FORWARD -p tcp -d " + dnp3rf.getSrc_IP() + " -s " + dnp3rf.getDst_IP();
            string dnp3_rules_from_client_to_server_log = "iptables -A FORWARD -p tcp -s " + dnp3rf.getSrc_IP() + " -d " + dnp3rf.getDst_IP() + " --dport 20000 -m state --state ESTABLISHED -j LOG --log-prefix " + "\"" + "ACCEPT&DNP3&ESTABLISHED " + "\"";
            // String dnp3_rules_from_server_to_client_established = "iptables -A FORWARD -p tcp -s " + dnp3rf.getDst_IP() + " -d " + dnp3rf.getSrc_IP() + " --sport 20000 -m state --state ESTABLISHED -j ACCEPT";
            if (add_delete == true)
            {
                flag = "DPI1";
                string frule =devform.getDev_IP()+" "+ dnp3rf.getDst_IP() + " "+ dnp3rf.getSrc_IP() + " ACCEPT "+log_flag.ToString();
                of.SaveRules(frule, "dnp3");
            }

            else if (add_delete == false)
            {
                flag = "DPI0";
                string frule = devform.getDev_IP() + " " + dnp3rf.getDst_IP() + " " + dnp3rf.getSrc_IP() + " ACCEPT " + log_flag.ToString();
                of.DeleteRules(frule, "dnp3");
            }
            string rule = flag + dnp3_rules_from_client_to_server_log + " && " + dnp3_rules_from_client_to_server_new + " && " + dnp3_rules_from_client_to_server_established;
            SendInfo sendcmd = new SendInfo(devform);      
            return sendcmd.SendConfigInfo(rule);
        }

        public bool ClearAllRules()
        {
            string rule = "iptables -P FORWARD ACCEPT && iptables -F && iptables -X && iptables -Z && iptables-restore</etc/iptables.up.rules";

            SendInfo sendcmd = new SendInfo(devform);
            return sendcmd.SendConfigInfo(rule);
            //sendcmd.SendConfigInfo("kill 'ps -e | grep snort | awk '{print $1}' |head -1"+"!");
        }

    }
}
